### ISSUE TITLE
DevToolsActivePort file doesn't exist - ChromeDriver through Selenium on Ubuntu Server

### ISSUE THREAD

**Issue by colinrsmall** [SeleniumHQ/selenium#6049]
## Meta -
OS:  Ubuntu Server 16.04

Selenium Version:  Java, 3.12.0

Browser:  ChromeDriver 2.37 - 2.40

Browser Version:  

## Expected Behavior -

Program starts a ChromeDriver succesfully:
`WebDriver gamePageDriver = new ChromeDriver(chromeOptions);`

## Actual Behavior -

`Exception in thread "main" org.openqa.selenium.WebDriverException: unknown error: DevToolsActivePort file doesn't exist
  (Driver info: chromedriver=2.40.565383 (76257d1ab79276b2d53ee976b2c3e3b9f335cde7),platform=Linux 4.4.0-127-generic x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 60.12 seconds
Build info: version: '3.12.0', revision: '7c6e0b3', time: '2018-05-08T15:15:03.216Z'
System info: host: 'vultr.guest', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-127-generic', java.version: '1.8.0_171'
Driver info: driver.version: ChromeDriver
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:214)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:166)
	at org.openqa.selenium.remote.JsonWireProtocolResponse.lambda$new$0(JsonWireProtocolResponse.java:53)
	at org.openqa.selenium.remote.JsonWireProtocolResponse.lambda$getResponseFunction$2(JsonWireProtocolResponse.java:91)
	at org.openqa.selenium.remote.ProtocolHandshake.lambda$createSession$0(ProtocolHandshake.java:123)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:126)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:73)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:136)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:83)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:543)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:207)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:130)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:181)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:168)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:157)
	at colinrsmall.webscraper.WebScraper.scrape(WebScraper.java:262)
	at colinrsmall.webscraper.WebScraper.main(WebScraper.java:54)`

## Steps to reproduce -

Following most steps here: https://fabianlee.org/2017/01/17/selenium-running-headless-automated-tests-on-ubuntu/

On a fresh Ubuntu Server 16.04 install:
`sudo apt-get install git openjdk-8-jdk maven -y
wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | sudo apt-key add -
sudo sh -c 'echo deb http://dl.google.com/linux/chrome/deb/ stable main > /etc/apt/sources.list.d/google.list'
sudo apt-get update
sudo apt-get install -y xfonts-100dpi xfonts-75dpi xfonts-scalable xfonts-cyrillic xvfb x11-apps imagemagick firefox google-chrome-stable'

Copy latest chromedriver_linux64 version to convenient location

In a separate SSH session: 
'/usr/binXvfb :99`

Run any .jar creating a new ChromeDriver. In my case, I am using the --headless and --disable-dev-shm-usage (as suggested by previous issues on the topic) options, but the provided example in the link above uses no options and also fails to start.

**Comment by rbclark** [comment]
Could you elaborate on this? As far as I can tell, the behavior I am seeing appears when we move from selenium 3.12.0 to 3.13.0, without changing the backing chromium version. This would indicate to me at least that the problem stems from a change in selenium, not in chromium.

**Comment by vanquy72dn** [comment]
The issue still happens on my system: Linux Ubuntu 16
{noformat}
2019-03-14 10:39:25,305 ERROR [Log] Unexpected exception while setting browser driver: org.openqa.selenium.WebDriverException: unknown error: Chrome failed to start: exited abnormally
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /var/lib/jenkins/.m2/repository/webdriver/chromedriver/linux64/2.45/chromedriver is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
  (Driver info: chromedriver=2.45.615279 (12b89733300bd268cff3b78fc76cb8f3a7cc44e5),platform=Linux 4.4.0-142-generic x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 70 milliseconds
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'ubuntudev', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-142-generic', java.version: '1.8.0_191'

{noformat}
Please help me to resolve it.

**Comment by cgoldberg** [comment]
please submit ChromeDriver issues to the Chromium team:
https://bugs.chromium.org/p/chromedriver

**Comment by eespitiarea** [comment]
That has been a major headache for me and invested some time in tracking the issue, is originated on Chromedriver side. A workaround is to specify the remote debugging port as an argument for your ChromeOptions, something like: `options.add_argument("--remote-debugging-port=9222")`, and if you're running in a container make the driver headless like so `options.headless = True`, otherwise use something like pyvirtualdisplay for screenshots.

Also, as a personal suggestion, if dealing with custom Chromium-like binaries, use Operadriver like so: `driver = webdriver.Chrome("operadriver", options=options)`, you won't run into so many issues regarding versions.

Hope that's useful @rpq.

**Comment by Chazza** [comment]
> --remote-debugging-port=9222

After doing that, I get:
```
Invalid --log-level value.
Unable to initialize logging. Exiting...
Exception in thread "main" org.openqa.selenium.WebDriverException: unknown error: Chrome failed to start: exited abnormally
  (chrome not reachable)
```

Any way around this?

Thanks

**Comment by eespitiarea** [comment]
Could you give me a bit more info? Like what options are you using, is this docker, what's your OS?

**Comment by saltypatrick** [comment]
Ran it as root in my case and the issue was that my terminal session was root and executing as root, but the desktop session was for my own non sudo user.

I ran it as my own server user nonroot and it worked just fine.
I noticed the same behavior in a containerized environment.
**TLDR/simple answer for now:** do not run chrome as root. In any environment you really shouldn't be doing this anyways, and I'd go as far as chrome does not support it (chrome or chromium binaries) as both issue out warnings.

**Comment by nekrozon** [comment]
> Ran it as root in my case and the issue was that my terminal session was root and executing as root, but the desktop session was for my own non sudo user.
> 
> I ran it as my own server user nonroot and it worked just fine.
> I noticed the same behavior in a containerized environment.
> **TLDR/simple answer for now:** do not run chrome as root. In any environment you really shouldn't be doing this anyways, and I'd go as far as chrome does not support it (chrome or chromium binaries) as both issue out warnings.

Yep.  I had a leftover sudo -s causing me this issue ... for over an hour... THANK YOU!
